### PR TITLE
add set_format to DatasetDict + tests

### PR DESCRIPTION
**Pull request by thomwolf**
Add the `set_format` and `formated_as` and `reset_format` to `DatasetDict`.
Add tests to these for `Dataset` and `DatasetDict`.
Fix some bugs uncovered by the tests for `pandas` formating.

### CODE REVIEW

File: tests/test_arrow_dataset.py [hunk 1]
```diff
@@ -11,12 +11,18 @@
...
-            pa.Table.from_pydict({"filename": ["my_name-train" + "_" + str(x) for x in np.arange(30).tolist()]})
-        )
+    def _create_dummy_dataset(self, multiple_columns=False):
+        if multiple_columns:
+            data = {"col_1": [3, 2, 1, 0], "col_2": ["a", "b", "c", "d"]}
+            dset = Dataset.from_dict(data)
+        else:
+            dset = Dataset(
+                pa.Table.from_pydict({"filename": ["my_name-train" + "_" + str(x) for x in np.arange(30).tolist()]})
+            )
```
Comment (lhoestq) [inline tests/test_arrow_dataset.py:25]: I guess we can use Dataset.from_dict here too

File: tests/test_dataset_dict.py [hunk 2]
```diff
@@ -4,21 +4,147 @@
...
+
+            cache_file_names = {
+                "train": os.path.join(tmp_dir, "train.arrow"),
+                "test": os.path.join(tmp_dir, "test.arrow"),
+            }
+            mapped_dsets_2: Dict[str, Dataset] = mapped_dsets_1.map(
+                lambda ex: {"bar": ["foo"] * len(ex["filename"])}, batched=True, cache_file_names=cache_file_names
+            )
+            self.assertListEqual(list(dsets.keys()), list(mapped_dsets_2.keys()))
+            self.assertListEqual(mapped_dsets_2["train"].column_names, ["filename", "foo", "bar"])
```
Comment (lhoestq) [inline tests/test_dataset_dict.py:147]: This test doesn't seem to test `set_format`

File: src/nlp/dataset_dict.py [hunk 3]
```diff
@@ -16,6 +17,71 @@ def _check_values_type(self):
                     "Values in `DatasetDict` should of type `Dataset` but got type '{}'".format(type(dataset))
                 )
 
+    @contextlib.contextmanager
+    def formated_as(
```
Comment (lhoestq) [inline src/nlp/dataset_dict.py:21]: You can add these methods to the docs in `docs/source/package_reference/main_classes.rst`

File: tests/test_dataset_dict.py (hunk 2 above)
Comment (thomwolf) [reply tests/test_dataset_dict.py:147]: Oh yes!

File: src/nlp/dataset_dict.py (hunk 3 above)
Comment (thomwolf) [reply src/nlp/dataset_dict.py:21]: 😄

**Pull request merged**